botocore/endpoints_v2.py

```diff
@@ -0,0 +1,1123 @@
+
+    def is_func(self, argument):
+        """Determine if an object is a function object.
+
+        :type argument: Any
+        :param argument: An object that may or may not be a function object.
+
+        :returns: A bool indicating if the input is a function object.
+        """
+        return isinstance(argument, dict) and "fn" in argument
+
+    def is_ref(self, argument):
+        """Determine if an object is a reference object.
+
+        :type argument: Any
+        :param argument: An object that may or may not be a reference object.
+
+        :returns: A bool indicating if the input is a reference object.
+        """
+        return isinstance(argument, dict) and "ref" in argument
+
+    def is_template(self, argument):
+        """Determine if an object contains a template string.
+
+        :type argument: Any
+        :param argument: An object that may or may not contain a template string.
+
+        :returns: A bool indicating if the input contains a template string.
+        """
+        return (
+            isinstance(argument, str)
+            and TEMPLATE_STRING_RE.search(argument) is not None
+        )
+
+    def resolve_template_string(self, value, scope_vars):
+        """Given a templated string, parse out the template nested between curly
+        braces, and return the nested data from the rule's scoped variables. If the
+        string contains double curly braces such as `{{foo}}` they are treated as
+        the string literal `{foo}` and are not formatted.
+
+        :type value: str
+        :param value: A template string such as "https://{Bucket}.ec2.{url#authority}".
+
+        :type scope_vars: dict
+        :param scope_vars: A dictionary combining input parameters and variables that
+        have been scoped throughout the rule's evaluation.
+
+        :returns: A fully resolved string such as "https://us-west-2.ec2.amazonaws.com".
+        """
+        string_formatter = Formatter()
+        result = ""
+        for literal, reference, _, _ in string_formatter.parse(value):
+            if reference is not None:
+                template_value = scope_vars
+                template_params = reference.split("#")
+                for param in template_params:
+                    template_value = template_value[param]
+                result += f"{literal}{template_value}"
+            else:
+                result += literal
+        return result
+
+    def convert_func_name(self, value):
+        """Given a function signature name found in the `fn` key of a function
+        object, convert it to snake case with a leading underscore. Additionally
+        replace a period `.` with an underscore `_`. For example, `aws.partition`
+        becomes `_aws_partition` and `booleanEquals` becomes `_boolean_equals`.
+
+        :type value: str
+        :param value: A string value representing the function name.
+
+        :returns: The actual name of a function defined in this module.
+        """
```

Comment:
Another note on docstrings, if there's more explanation than code, that's often a bad smell. There are exceptions when there's more intricate nuance, but as a general rule I'd aim for a succinct sentence (or two) in most cases. The code should tell most of the story and we don't have to worry about docstring/functionality drift over time.

Reply:
Sure. I lifted a lot of this text from the spec, but made no effort to consolidate. Will do.